### PR TITLE
JS API to open card browser and search with query 

### LINKED ISSUE #8109: js api -> ability to open card browser (the card search) from js api with `ankiSearch("deck:foo myField:bar")`
**Describe alternatives you've considered**

related to:
- https://faqs.ankiweb.net/linking-cards-together.html
- https://ankiweb.net/shared/info/1423933177

the current solution to make links from card to card is to:
- render all cards as html files
- place them in collection.media folder on mobile or pc
- add Iframes `<iframe src="_card${cid}.html" width="${0.9*window.innerWidth}" height="${0.9*window.innerHeight}" ></iframe> `

but you cannot edit the related card this way

**Is your feature request related to a problem? Please describe.**

But for me the ability to issue the search request / open card browser would be sufficient

**Describe the solution you'd like**

in back of my card I would have sentence

```
<div onclick="ankiSearch("deck:\"My Deck\" kanji:kanji1")">kanji1</div>
```

and on press it would open card browser (without closing the card!!!, I can still return to it by pressing back button)

and fill it with search query

the search query parser would extract the:
1. "deck:..." - if no deck is in query THEN use "All Decks"
2. (maybe) [`is:suspended`](https://github.com/ankidroid/Anki-Android/blob/c2f6c1080a099800faaf83e2690abd7e343d7d1b/AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java#L1072)
3. (maybe) `tag:marked`
4. [...all other fields are considered as query to `setQuery`](https://github.com/ankidroid/Anki-Android/blob/c2f6c1080a099800faaf83e2690abd7e343d7d1b/AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java#L1073)

**Additional context**

photo5280945011405468666

### PR DESCRIPTION
**Pull request by infinyte7**
## Purpose / Description
This is continuation of JS API features. From reviewer using js function/button in one click it will go to CardBrowser and search for cards using query.

## Fixes
Fixes #8109

## Approach
1. Added a JSInterface function and used intent to call CardBrowser
3. Get string extra in CardBrowser and assign query to `mSearchTerms` then call `searchCards()`

## How Has This Been Tested?
Testesd on devices and emulator 

1. Add a button in card template to call `AnkiDroidJS.ankiSearchCard(query)`
2. Add query like, this query are same as entering in CardBrowser
```
'deck:"deck name" {{front}}'
'deck:"test" flag:2 tag:marked'
'deck:"test" flag:2 tag:marked is:suspended'
'deck:"test" flag:2 tag:marked search text here'
```
3. Click button to run the query
```js
<button onclick="AnkiDroidJS.ankiSearchCard('deck:\"test\" {{hanzi}}')">Search this in Card browser</button>
```

### Note in this video: Add `AnkiDroidJS.` before `ankiSearchCard` because JSInterface used to call this function.
https://user-images.githubusercontent.com/12841290/125199042-0bbff280-e297-11eb-9670-07087c713b96.mp4

#### Another example, Added following to front side of card template
```js
<button onclick="searchCard();">{{Simplified}}</button>     //  <-----  note this line

<script>
function searchCard() {
AnkiDroidJS.ankiSearchCard('deck:"xiehanzi HSK 3.0" {{Simplified}}');  //  <-----  note this line
}
</script>

<div class="text-color1">{{Pinyin}}</div>
<div class="text-color3">{{Meaning}}</div>
```

#### image
</img>

## Learning (optional, can help others)
https://github.com/ankidroid/Anki-Android/wiki/AnkiDroid-Javascript-API

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java [hunk 1]
```diff
@@ -597,6 +597,11 @@ protected void onCreate(Bundle savedInstanceState) {
         setContentView(R.layout.card_browser);
         initNavigationDrawer(findViewById(android.R.id.content));
         startLoadingCollection();
+
+        if (getIntent().getBooleanExtra("called_from_js_api", false)) {
+            mSearchTerms = getIntent().getStringExtra("search_query");
+            searchCards();
+        }
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:604]: If you just checked for the search_query Intent extra, and used it if non-null, then all the "called_from_js_api" stuff can disappear, making this a much smaller / easier to understand change

```suggestion

        if (getIntent().getStringExtra("search_query") != null) {
            mSearchTerms = getIntent().getStringExtra("search_query");
            searchCards();
        }
```

File: AnkiDroid/src/main/java/com/ichi2/anki/NavigationDrawerActivity.java [hunk 2]
```diff
@@ -418,6 +418,18 @@ protected void openCardBrowser() {
...
+    }
+
+    protected void openCardBrowser(String query) {
+        Intent intent = new Intent(NavigationDrawerActivity.this, CardBrowser.class);
+        Long currentCardId = getCurrentCardId();
+        if (currentCardId != null) {
+            intent.putExtra("currentCard", currentCardId);
+        }
+        intent.putExtra("called_from_js_api", true);
+        intent.putExtra("search_query", query);
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/NavigationDrawerActivity.java:432]: Change the official "openCardBrowser" API to add the query parameter as a `@Nullable` item and update all callers to the new API

Do this as a separate clean commit, then review of that change is trivial

As a second commit then add the Javascript API part that uses it - then this looks good to go?

```suggestion
    protected void openCardBrowser(String query) {
        Intent intent = new Intent(NavigationDrawerActivity.this, CardBrowser.class);
        Long currentCardId = getCurrentCardId();
        if (currentCardId != null) {
            intent.putExtra("currentCard", currentCardId);
        }
        intent.putExtra("search_query", query);
```

**Review by mikehardy — CHANGES_REQUESTED** [review]
I like this idea, but as mentioned in comments, split into 2 commits where one alters the API and all callers to always use Nullable query string to open card browser as a small simple change, then second commit layers the new Javascript API on top of it using the new capability

**Comment by infinyte7** [comment]
> I like this idea, but as mentioned in comments, split into 2 commits where one alters the API and all callers to always use Nullable query string to open card browser as a small simple change, then second commit layers the new Javascript API on top of it using the new capability

I will split into two PR. Thanks

**Comment by mikehardy** [comment]
For this one could just be 2 commits I think - in this same PR, that would work for me, these ones will be small

**Comment by infinyte7** [comment]
> For this one could just be 2 commits I think - in this same PR, that would work for me, these ones will be small

Okay, I will do it.

**Review by mikehardy — CHANGES_REQUESTED** [review]
Can you use `@JavascriptInterface` in this one as well? Like in #9245 ?

**Review by mikehardy — APPROVED** [review]
Nice!

**Comment by mikehardy** [comment]
Has a conflict with your other PR though :sweat_smile:

**Comment by infinyte7** [comment]
Yes, I think I need to rebase it.

I also thinking about implementing in signal. Because here are two types of function
- Functions that get parameter(s) - `ankiShowToast, ankiMarkCard etc.`
- Functions that do not get parameter(s) -  `All JSInterface functions`

So, should we make a rule that function that needs parameter(s) should go through signal and the function which do no require parameter(s) should go though `JSInterface`?

**Comment by mikehardy** [comment]
- I am not aware of any security concerns for JavascriptInterface anymore (android 5+). Arguments, no arguments, no matter.
- I am under the impression *everything* should be JavascriptInterface, I expect this to be JavascriptInterface
- The commit history here is not good. There is no reason to have more than one commit here, containing nothing but the commit that adds the JavascriptInterface. Rebase is what you need to get there

**Comment by infinyte7** [comment]
Okay, I am on it.

**Review by mikehardy — APPROVED** [review]
Nice!

**Pull request merged**